### PR TITLE
Add TS support to fields

### PR DESCRIPTION
**Pull request by djhi**
Follow #8862 

If users provide types **extending** `RaRecord`, nothing changes, they won't have TS validation of the `source` prop but they won't have any error either. Same when they don't provide types.

If users provide types that don't extend `RaRecord`, they'll get validation and completion on the `source` prop

### CODE REVIEW

**Review by fzaninotto — CHANGES_REQUESTED** [review]

File: examples/crm/src/types.ts [hunk 1]
```diff
@@ -1,12 +1,12 @@
-import { RaRecord, Identifier } from 'react-admin';
-
-export interface Sale extends RaRecord {
+export type Sale = {
```
[the remark this answers is not in the document]
Comment (djhi) [reply examples/crm/src/types.ts]: Reverted

File: docs/Fields.md [hunk 2]
```diff
@@ -597,4 +597,34 @@ You can find components for react-admin in third-party repositories.
...
+}
+
+export const PostShow = () => (
+    <Show>
+        <SimpleShowLayout>
+            <TextField<Post> source="title" />
+            <TextField<Post> source="teaser" />
+            <RichTextField<Post> source="body" />
+            <DateField<Post> label="Publication date" source="published_at" />
+        </SimpleShowLayout>
```
[the remark this answers is not in the document]
Comment (djhi) [reply docs/Fields.md:628]: Added

File: examples/demo/src/types.ts [hunk 3]
```diff
@@ -1,13 +1,13 @@
-import { RaRecord, Identifier } from 'react-admin';
-
 export type ThemeName = 'light' | 'dark';
 
-export interface Category extends RaRecord {
+export type Category = {
```
[the remark this answers is not in the document]
Comment (djhi) [reply examples/demo/src/types.ts]: Reverted

File: examples/demo/src/visitors/ColoredNumberField.tsx [hunk 4]
```diff
@@ -13,6 +13,8 @@ const ColoredNumberField = (props: NumberFieldProps) => {
     );
 };
 
-ColoredNumberField.defaultProps = NumberField.defaultProps;
+ColoredNumberField.defaultProps = {
```
[the remark this answers is not in the document]
Comment (djhi) [reply examples/demo/src/visitors/ColoredNumberField.tsx]: Reverted

File: packages/ra-ui-materialui/src/field/ArrayField.tsx [hunk 5]
```diff
@@ -74,30 +75,37 @@ import { PublicFieldProps, InjectedFieldProps, fieldPropTypes } from './types';
...
-export const ArrayField = memo<ArrayFieldProps>(props => {
+const ArrayFieldImpl = <
+    RecordType extends Record<string, unknown> = Record<string, any>
+>(
+    props: ArrayFieldProps<RecordType>
+) => {
     const { children, resource, source, perPage, sort, filter } = props;
     const record = useRecordContext(props);
-    const data = get(record, source, emptyArray) || emptyArray;
+    const data = (get(record, source, emptyArray) as any[]) || emptyArray;
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/field/ArrayField.tsx:85]: `Array<Record<string, any>>` would be better than `any[]`

File: packages/ra-ui-materialui/src/field/ArrayField.tsx [hunk 6]
```diff
@@ -74,30 +75,37 @@ import { PublicFieldProps, InjectedFieldProps, fieldPropTypes } from './types';
...
-
-// @ts-ignore
-ArrayField.propTypes = {
-    ...fieldPropTypes,
 };
 
-export interface ArrayFieldProps extends PublicFieldProps, InjectedFieldProps {
+export const ArrayField = genericMemo(ArrayFieldImpl);
+
+ArrayFieldImpl.propTypes = fieldPropTypes;
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/field/ArrayField.tsx:96]: I'd expect this line to come before the memo

File: packages/ra-ui-materialui/src/field/ArrayField.tsx [hunk 7]
```diff
@@ -74,30 +75,37 @@ import { PublicFieldProps, InjectedFieldProps, fieldPropTypes } from './types';
...
+export interface ArrayFieldProps<
+    RecordType extends Record<string, unknown> = Record<string, any>
+> extends FieldProps<RecordType> {
     children: ReactNode;
     perPage?: number;
     sort?: SortPayload;
     filter?: FilterPayload;
 }
 
+ArrayFieldImpl.displayName = 'ArrayFieldImpl';
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/field/ArrayField.tsx:108]: same

File: packages/ra-ui-materialui/src/field/BooleanField.tsx [hunk 8]
```diff
@@ -85,11 +88,14 @@ BooleanField.propTypes = {
     looseValue: PropTypes.bool,
 };
 
+BooleanField.propTypes = BooleanFieldImpl.propTypes;
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/field/BooleanField.tsx:91]: Same, please define the proptypes of the implementation first then copy it to the memoized component. By the way, can't you bake in this proptype copy in the `genericMemo` component? Same for the `displayName` copy.

File: packages/ra-ui-materialui/src/field/BooleanField.stories.tsx [hunk 9]
```diff
@@ -22,3 +22,32 @@ export const Basic = (props: BooleanFieldProps) => {
 export const NoFalseIcon = () => <Basic FalseIcon={null} />;
 
 export const NoTrueIcon = () => <Basic TrueIcon={null} />;
+
+type Post = {
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/field/BooleanField.stories.tsx:26]: That's not a record, it requires an `id`. I'm surprised that TypeScript doesn't warn you about this one.

File: packages/ra-ui-materialui/src/field/BooleanField.stories.tsx [hunk 10]
```diff
@@ -22,3 +22,32 @@ export const Basic = (props: BooleanFieldProps) => {
...
+            <input
+                type="checkbox"
+                checked={published}
+                onChange={e => setPublished(e.target.checked)}
+            />
+            <BooleanField<Post>
+                record={{ published, deep: { reported: false }, title: '' }}
+                source="deep.reported"
+                sortBy="published"
+                {...props}
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/field/BooleanField.stories.tsx:49]: Why do you need that? It shows a practice that shouldn't be encouraged.

File: packages/ra-ui-materialui/src/field/DateField.tsx [hunk 11]
```diff
@@ -70,7 +74,13 @@ export const DateField: FC<DateFieldProps> = memo(props => {
         ) : null;
     }
 
-    const date = value instanceof Date ? value : new Date(value);
+    const date =
+        value instanceof Date
+            ? value
+            : typeof value === 'string' || typeof value === 'number'
+            ? new Date(value)
+            : undefined;
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/field/DateField.tsx:82]: 👍 TypeScript is useful after all!

File: packages/ra-ui-materialui/src/field/ReferenceField.tsx [hunk 12]
```diff
@@ -103,54 +109,48 @@ ReferenceField.defaultProps = {
...
-    const createPath = useCreatePath();
-    const resourceDefinition = useResourceDefinition({ resource: reference });
-
-    const resourceLinkPath =
-        link === false ||
-        (link === 'edit' && !resourceDefinition.hasEdit) ||
-        (link === 'show' && !resourceDefinition.hasShow)
-            ? false
-            : createPath({
-                  resource: reference,
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/field/ReferenceField.tsx:135]: Any particular reason you moved that logic down to PureReferenceFieldView? By the way that's a BC, as ReferenceFieldView is public.

File: packages/ra-ui-materialui/src/field/ReferenceOneField.tsx [hunk 13]
```diff
@@ -84,7 +87,7 @@ export const ReferenceOneField = <RecordType extends RaRecord = any>(
                 isLoading={isLoading}
                 isFetching={isFetching}
                 referenceRecord={referenceRecord}
-                resourceLinkPath={resourceLinkPath}
+                link={resourceLinkPath}
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/field/ReferenceOneField.tsx:90]: If you had to change it here, it will break other components (e.g. Enterprise components), so you should revert this change.

File: packages/ra-ui-materialui/src/field/FileField.tsx [hunk 14]
```diff
@@ -37,7 +41,7 @@ export const FileField = (props: FileFieldProps) => {
         ...rest
     } = props;
     const record = useRecordContext(props);
-    const sourceValue = get(record, source);
+    const sourceValue = get(record, source) as string;
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/field/FileField.tsx:44]: Why don't you use `toString()` as in `RichTextField` and `TextField`? Seems a better way to make sure that what you get is a string. 

Same for other `as string` forced casting.

**Review by fzaninotto — CHANGES_REQUESTED** [review]
Awesome!

File: packages/ra-core/src/types.ts [hunk 15]
```diff
@@ -14,8 +13,8 @@ export interface RaRecord<IdentifierType extends Identifier = Identifier>
     id: IdentifierType;
 }
 
-export interface SortPayload {
-    field: string;
+export interface SortPayload<FieldType extends string = string> {
+    field: FieldType;
```
[the remark this answers is not in the document]
Comment (djhi) [reply packages/ra-core/src/types.ts:17]: I'll remove that but later, users could provide their type and by default, this will match the resource properties as well, just like `source`

File: packages/ra-ui-materialui/package.json [hunk 16]
```diff
@@ -62,6 +62,7 @@
         "clsx": "^1.1.1",
         "css-mediaquery": "^0.1.2",
         "dompurify": "^2.4.3",
+        "hotscript": "^1.0.12",
```
[the remark this answers is not in the document]
Comment (djhi) [reply packages/ra-ui-materialui/package.json:65]: Because **you** asked me to put it in dependencies ^^
image

File: packages/ra-ui-materialui/src/button/SaveButton.tsx [hunk 17]
```diff
@@ -151,7 +151,7 @@ interface Props<
     mutationOptions?: UseMutationOptions<
         RecordType,
         MutationOptionsError,
-        CreateParams<RecordType> | UpdateParams<RecordType>
+        CreateParams | UpdateParams<RecordType>
```
[the remark this answers is not in the document]
Comment (djhi) [reply packages/ra-ui-materialui/src/button/SaveButton.tsx:154]: Can revert for now but related to our discussion about the RecordType and the FormValuesType

File: packages/ra-ui-materialui/src/field/types.ts [hunk 18]
```diff
@@ -1,19 +1,48 @@
...
+    emptyText?: string;
+    fullWidth?: boolean;
+    record?: RecordType;
+    resource?: string;
+}
 
-export interface PublicFieldProps {
-    sortBy?: string;
+/**
+ * @deprecated use FieldProps instead
```
[the remark this answers is not in the document]
Comment (djhi) [reply packages/ra-ui-materialui/src/field/types.ts:35]: Because it's a duplicate of `Fields` that has no purpose. Why keep it?

File: packages/ra-ui-materialui/src/field/BooleanField.stories.tsx (hunk 9 above)
Comment (djhi) [reply packages/ra-ui-materialui/src/field/BooleanField.stories.tsx:26]: Because our fields don't require a RaRecord anymore

File: packages/ra-ui-materialui/src/field/ArrayField.tsx [hunk 19]
```diff
@@ -74,30 +75,35 @@ import { PublicFieldProps, InjectedFieldProps, fieldPropTypes } from './types';
...
             {children}
         </ListContextProvider>
     );
-});
-
-// @ts-ignore
-ArrayField.propTypes = {
-    ...fieldPropTypes,
 };
+ArrayFieldImpl.propTypes = fieldPropTypes;
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/field/ArrayField.tsx:95]: We used to clone the prop types before. There was probably a reason, I think you shouldn't change that.

File: examples/crm/src/contacts/TagsListEdit.tsx [hunk 20]
```diff
@@ -61,18 +61,16 @@ export const TagsListEdit = () => {
...
         update('contacts', {
             id: record.id,
             data: { tags },
             previousData: record,
         });
     };
 
-    const handleAddTag = (id: Identifier) => {
-        const tags: Identifier[] = [...record.tags, id];
+    const handleAddTag = (id: number) => {
```
[the remark this answers is not in the document]
Comment (fzaninotto) [reply examples/crm/src/contacts/TagsListEdit.tsx]: Bump

File: packages/ra-core/src/dataProvider/withLifecycleCallbacks.ts
[the remark this answers is not in the document]
Comment (fzaninotto) [reply packages/ra-core/src/dataProvider/withLifecycleCallbacks.ts:1]: bump

File: packages/ra-ui-materialui/package.json (hunk 16 above)
[the remark this answers is not in the document]
Comment (fzaninotto) [reply packages/ra-ui-materialui/package.json:65]: bump

**Review by fzaninotto — CHANGES_REQUESTED** [review]

File: packages/ra-ui-materialui/src/field/ArrayField.tsx (hunk 19 above)
Comment (djhi) [reply packages/ra-ui-materialui/src/field/ArrayField.tsx:95]: I can't think of a valid reason. We do the same in EmailField.

I think someone did not simplify the syntax when removing an old propType specific to the ArrayField that did not make sense in v4. See https://github.com/marmelab/react-admin/commit/f7d9f18505996a3af92f29b31d3bc1d104052dfe

Comment (fzaninotto) [reply packages/ra-ui-materialui/src/field/ArrayField.tsx:95]: I vaguely remember having to clone proptypes for an obscure reason. Let's revert it if we have no reason to do it.

Comment (djhi) [reply packages/ra-ui-materialui/src/field/ArrayField.tsx:95]: What about the EmailField, UrlField and WrapperField then?

Comment (fzaninotto) [reply packages/ra-ui-materialui/src/field/ArrayField.tsx:95]: Let's keep them as in master. I know, in master there is no consistency. Let's keep it like that. 

My opinion is that this PR changes quite a lot of stuff, and we should avoid all unnecessary changes to reduce any potential BC break.

**Review by fzaninotto — APPROVED** [review]

**Comment by akillingbeck89** [comment]
Just upgraded from 4.5.1 to 4.11.1 and having TS issues all over the place where I set my "source" parameter, is there a way to disable this?

**Comment by slax57** [comment]
@akillingbeck89 This change should be backwards compatible. If it is not, this is to be considered as a bug.
Can you please open a new issue with more information about the errors you are getting, and ideally a reproduction sandbox?
This way we can better grasp what's going on and offer adequate help.
For instance, it would help to know if you are passing custom types to the fields or not.

**Comment by grevolution** [comment]
> Dont have time right now, will do it next week, but no custom types are being passed and an example error I'm getting is:
> 
> Type '"channelId"' is not assignable to type '`[${key}]` | `[${key}].${string}` | null'.ts(2322)

How did you fix this issue?
